File: 04/claude-opus-4.5/go/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	data, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	var grid []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if len(line) > 0 {
			grid = append(grid, line)
		}
	}

	rows := len(grid)
	cols := len(grid[0])

	// Part 1: Count rolls with fewer than 4 adjacent rolls
	dx := []int{-1, -1, -1, 0, 0, 1, 1, 1}
	dy := []int{-1, 0, 1, -1, 1, -1, 0, 1}

	var part1 int64 = 0
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] != '@' {
				continue
			}

			adjacentRolls := 0
			for d := 0; d < 8; d++ {
				nr := r + dx[d]
				nc := c + dy[d]
				if nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == '@' {
					adjacentRolls++
				}
			}

			if adjacentRolls < 4 {
				part1++
			}
		}
	}

	fmt.Printf("Day 04 Part 1: %d\n", part1)

	// Part 2: Iteratively remove accessible rolls until none remain
	mutableGrid := make([][]byte, rows)
	for i := range grid {
		mutableGrid[i] = []byte(grid[i])
	}

	var part2 int64 = 0

	for {
		var toRemove [][2]int

		for r := 0; r < rows; r++ {
			for c := 0; c < cols; c++ {
				if mutableGrid[r][c] != '@' {
					continue
				}

				adjacentRolls := 0
				for d := 0; d < 8; d++ {
					nr := r + dx[d]
					nc := c + dy[d]
					if nr >= 0 && nr < rows && nc >= 0 && nc < cols && mutableGrid[nr][nc] == '@' {
						adjacentRolls++
					}
				}

				if adjacentRolls < 4 {
					toRemove = append(toRemove, [2]int{r, c})
				}
			}
		}

		if len(toRemove) == 0 {
			break
		}

		for _, pos := range toRemove {
			mutableGrid[pos[0]][pos[1]] = '.'
		}

		part2 += int64(len(toRemove))
	}

	fmt.Printf("Day 04 Part 2: %d\n", part2)
}

File: 04/claude-opus-4.5/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input");
    let grid: Vec<&str> = input.trim().lines().filter(|l| !l.is_empty()).collect();
    let rows = grid.len();
    let cols = grid[0].len();

    // Part 1: Count rolls with fewer than 4 adjacent rolls
    let dx: [i32; 8] = [-1, -1, -1, 0, 0, 1, 1, 1];
    let dy: [i32; 8] = [-1, 0, 1, -1, 1, -1, 0, 1];

    let mut part1: i64 = 0;
    for r in 0..rows {
        for c in 0..cols {
            if grid[r].chars().nth(c).unwrap() != '@' {
                continue;
            }

            let mut adjacent_rolls = 0;
            for d in 0..8 {
                let nr = r as i32 + dx[d];
                let nc = c as i32 + dy[d];
                if nr >= 0 && nr < rows as i32 && nc >= 0 && nc < cols as i32 {
                    if grid[nr as usize].chars().nth(nc as usize).unwrap() == '@' {
                        adjacent_rolls += 1;
                    }
                }
            }

            if adjacent_rolls < 4 {
                part1 += 1;
            }
        }
    }

    println!("Day 04 Part 1: {}", part1);

    // Part 2: Iteratively remove accessible rolls until none remain
    let mut mutable_grid: Vec<Vec<char>> = grid.iter().map(|row| row.chars().collect()).collect();

    let mut part2: i64 = 0;

    loop {
        let mut to_remove: Vec<(usize, usize)> = Vec::new();

        for r in 0..rows {
            for c in 0..cols {
                if mutable_grid[r][c] != '@' {
                    continue;
                }

                let mut adjacent_rolls = 0;
                for d in 0..8 {
                    let nr = r as i32 + dx[d];
                    let nc = c as i32 + dy[d];
                    if nr >= 0 && nr < rows as i32 && nc >= 0 && nc < cols as i32 {
                        if mutable_grid[nr as usize][nc as usize] == '@' {
                            adjacent_rolls += 1;
                        }
                    }
                }

                if adjacent_rolls < 4 {
                    to_remove.push((r, c));
                }
            }
        }

        if to_remove.is_empty() {
            break;
        }

        for (r, c) in &to_remove {
            mutable_grid[*r][*c] = '.';
        }

        part2 += to_remove.len() as i64;
    }

    println!("Day 04 Part 2: {}", part2);
}
